R/proxy.R

```diff
@@ -1,28 +1,54 @@
-#' Extract underlying data
+#' Proxy and restore
 #'
 #' @description
 #'
-#' Extract the data underlying an S3 vector object, i.e. the underlying
-#' (named) atomic vector or list.
+#' \Sexpr[results=rd, stage=render]{vctrs:::lifecycle("experimental")}
 #'
-#' * `vec_data()` returns unstructured data. The only attributes
-#'   preserved are names, dims, and dimnames.
+#' `vec_proxy()` returns the data structure containing the values of a
+#' vector. This data structure is usually the vector itself. In this
+#' case the proxy is the [identity function][base::identity], which is
+#' the default `vec_proxy()` method.
 #'
-#'   Currently, due to the underlying memory architecture of R, this
-#'   creates a full copy of the data.
+#' You normally don't need to implement `vec_proxy()`, except in these
+#' cases:
 #'
-#' * `vec_proxy()` may return structured data. This generic is the
-#'   main customisation point in vctrs, along with [vec_restore()].
-#'   See the section below to learn when you should implement
-#'   `vec_proxy()`.
+#' - Some vector types have vectorised attributes, i.e. metadata for
```

Comment:
```suggestion
#' - A vector has vectorised attributes, i.e. metadata for
```

Reply:
But maybe this bullet should be de-emphasised too, since it's only us that needs to implement this method?

Reply:
What do you mean it's only us? What if you have a POSIXlt class?

Reply:
Doesn’t the rcrd subclass take care of that for you?

Reply:
If you have a POSIXlt-like class you don't have a rcrd subclass.

The rcrd class is higher level than this documentation. In principle it can live in another package.

Reply:
I'll add a note that for classes started from scratch the rcrd class is the easiest way.

Reply:
Yeah, just make it clear that this is for experts only.

Reply:
By the way this made me wonder if we should make `fields()` more generic by defining it on any vector whose proxy is a data frame. Though overall I think these functions are not really needed and it feels a bit wrong to use up unnamespaced short names like this, especially since vctrs is meant to be imported.